### ISSUE TITLE
Idea: Networking/Sockets

### ISSUE THREAD

**Issue by isc30** [mosra/corrade#67]
It's really difficult to find a good/lightweight library with no dependencies that also compiles with emscripten regarding sockets.

* **SDL2_net** is supported but it adds a lot of overhead (it requires `-s USE_SDL=2 -s USE_SDL_NET=2`)
* **POCO** (latest version) doesn't compile well with emscripten
* **Boost.asio / Boost.Beast** are too complicated to use and I'm not sure if they are even supported by emscripten (maybe it is when you make it fallback to unix sockets)

I know this is a crazy idea that maybe should be added as a plugin/integration to Magnum but it would open a world of possibilities (Wasm + Graphics + Audio + Networking)

**Comment by mosra** [comment]
As far as I understand, you're building a WASM-only game/engine, so then something Emscripten-specific could work too, right? This introduction from @Squareys might help: https://blog.squareys.de/emscripten-sockets/ We actually discussed introducing wrappers for these APIs directly to Corrade/Magnum but then other things took a priority.

Looking at how Emscripten wraps websockets to mimic the Unix sockets, I'm wondering if there's a way to use those APIs directly, without the extra wrapper layer (and possibly code size / runtime overhead). ... actually, there seems to be something: https://emscripten.org/docs/api_reference/emscripten.h.html#socket-event-registration (I didn't use any of these myself, though).

Hope this helps!

**Comment by isc30** [comment]
Not exactly, my idea is to build a generic setup project that works well when targeting native mobile, web and desktop (client and server). This is why magnum is an excellent choice in terms of graphics and audio.

The ultimate goal would be to write a game I have in mind, but it requires a networking so I'm currently investigating what could I use. Ideally a thin abstraction layer over sockets would be enough but if it has more things (like a websocket protocol already implemented for the server) then it would be better.

I also discovered [HumbleNet](https://github.com/HumbleNet/HumbleNet) that looks like the best alternative for P2P games where the host could be one of the players.

I'm still thinking if making a new lib with custom optimized bindings in emscripten and part of the API using HumbleNet is a good idea.

**Comment by Squareys** [comment]
Hi @isc30 !

I am building a WASM-centered engine based on magnum and have written a networked WASM game with Magnum before ([=> WebVR Pong](https://constructarca.de/games/webvr-pong/), though the network mode is currently disabled):

 - Emscripten sockets are WebSockets, which is a TCP based protocoll and seemsto be not well suited for real-time networked games. Like HubleNet does, you should use WebRCT instead, probably, but I have not looked into that yet.
 - I have some `Corrade::Net::Socket` which has a Windows and Emscripten implementation. WebSockets are not normal sockets, as mentioned, hence on server side, there is a difference between an Emscripten-based client and a Windows-based client connecting.
 - WebSockets are not Peer-to-peer. WebRTC on the other hand is, afaik.
 - Poco is nice, I use it a lot for the server side. Since they implement the WebSockets protocol also, you can use that as a central game server when going the WebSockets route (like with WebVR pong). I would not recommend using it for an Emscripten game.

Abstractions are nice, but when it comes to sockets, trying to abstract those for emscripten seems to mainly cause a lot of confusion.

HumbeNet looks nice, using something like that is probably the way to go.

Feel free to ask more, if that doesn't answer your questions 🙂
